samples/bluetooth/peripheral_gatt_dm/src/main.c

```diff
@@ -160,8 +160,8 @@ static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
 	printk("Pairing failed conn: %s, reason %d\n", addr, reason);
 
 	if (pairing_confirmation_conn) {
-		pairing_confirmation_conn = NULL;
 		bt_conn_unref(pairing_confirmation_conn);
+		pairing_confirmation_conn = NULL;
```

Comment:
Is this order correct and safe?

You declare to the stack that a connection object is unrefed before it is actually unrefed.

Reply:
`pairing_confirmation_conn` is not declared atomic. So the change of order here does not theoretically change the thread safety of this code. The compiler is in fact allowed to reorder the NULL store and the call to `bt_conn_unref`. The code in the PR should have exactly the same meaning to an optimizing compiler as the following code.

```C
struct bt_conn *tmp = pairing_confirmation_conn;
pairing_confirmation_conn = NULL;
bt_conn_unref(tmp);
```

The compiler is likely not able to prove the correctness of this transformation, if we don't have link time optimization available, and so will not actually do it. But the compiler will also not emit any memory barriers necessary for synchronizing between threads. So relying on the code with `tmp` to actually commit `pairing_confirmation_conn = NULL` to memory and make it visible to other threads before the effect of `bt_conn_unref` is visible is not safe.

I have not verified that `pairing_confirmation_conn` is ok without atomics.

Reply:
I would say that referencing a conn object is  even not needed in this sample. Normal assigning pointer to the conn should be suitable here.